### ISSUE TITLE
SIGABRT with primitive type, new in 1.5

### ISSUE THREAD

**Issue by BioTurboNick** [JuliaLang/julia#37050]
I'm probably doing something very wrong here, but I presume that no matter how badly I'm doing things, Julia itself shouldn't get a SIGABRT?

This is the code that triggers the crash:
```
primitive type SalsaBlock <: Unsigned 512 end
SalsaBlock() = reinterpret(SalsaBlock, zeros(UInt128, 4))
Base.show(io::IO, x::SalsaBlock) = print(io, reinterpret(UInt128, x))
SalsaBlock()
```

In Julia 1.5.0, this code results in a SIGABRT:
```
Assertion failed: to != T_void, file /cygdrive/d/buildbot/worker/package_win64/build/src/intrinsics.cpp, line 327

signal (22): SIGABRT
in expression starting at none:0
crt_sig_handler at /cygdrive/d/buildbot/worker/package_win64/build/src\signals-win.c:92
raise at C:\WINDOWS\System32\msvcrt.dll (unknown line)
abort at C:\WINDOWS\System32\msvcrt.dll (unknown line)
assert at C:\WINDOWS\System32\msvcrt.dll (unknown line)
emit_unbox at /cygdrive/d/buildbot/worker/package_win64/build/src\intrinsics.cpp:327
emit_expr at /cygdrive/d/buildbot/worker/package_win64/build/src\codegen.cpp:3815
emit_ssaval_assign at /cygdrive/d/buildbot/worker/package_win64/build/src\codegen.cpp:3339
emit_stmtpos at /cygdrive/d/buildbot/worker/package_win64/build/src\codegen.cpp:3533 [inlined]
emit_function at /cygdrive/d/buildbot/worker/package_win64/build/src\codegen.cpp:6085
jl_emit_code at /cygdrive/d/buildbot/worker/package_win64/build/src\codegen.cpp:6431
jl_emit_codeinst at /cygdrive/d/buildbot/worker/package_win64/build/src\codegen.cpp:6465
jl_compile_workqueue at /cygdrive/d/buildbot/worker/package_win64/build/src\codegen.cpp:6568
_jl_compile_codeinst at /cygdrive/d/buildbot/worker/package_win64/build/src\jitlayers.cpp:100
jl_generate_fptr at /cygdrive/d/buildbot/worker/package_win64/build/src\jitlayers.cpp:302
jl_compile_method_internal at /cygdrive/d/buildbot/worker/package_win64/build/src\gf.c:1964
jl_compile_method_internal at /cygdrive/d/buildbot/worker/package_win64/build/src\gf.c:1931 [inlined]
_jl_invoke at /cygdrive/d/buildbot/worker/package_win64/build/src\gf.c:2224 [inlined]
jl_apply_generic at /cygdrive/d/buildbot/worker/package_win64/build/src\gf.c:2398
display at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\REPL\src\REPL.jl:214
display at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\REPL\src\REPL.jl:218
display at .\multimedia.jl:328
jl_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\julia.h:1690 [inlined]
do_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:655
jl_f__apply at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:669 [inlined]
jl_f__apply_latest at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:705
#invokelatest#1 at .\essentials.jl:710 [inlined]
invokelatest at .\essentials.jl:709 [inlined]
print_response at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\REPL\src\REPL.jl:238
print_response at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\REPL\src\REPL.jl:223
do_respond at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\REPL\src\REPL.jl:822
jfptr_do_respond_71821.clone_1 at C:\Users\nicho\AppData\Local\Programs\Julia 1.5.0\lib\julia\sys.dll (unknown line)
jl_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\julia.h:1690 [inlined]
do_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:655
jl_f__apply at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:669 [inlined]
jl_f__apply_latest at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:705
#invokelatest#1 at .\essentials.jl:710 [inlined]
invokelatest at .\essentials.jl:709 [inlined]
run_interface at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\REPL\src\LineEdit.jl:2355
jfptr_run_interface_70221.clone_1 at C:\Users\nicho\AppData\Local\Programs\Julia 1.5.0\lib\julia\sys.dll (unknown line)
run_frontend at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\REPL\src\REPL.jl:1143
#38 at .\task.jl:356
jfptr_YY.38_71712.clone_1 at C:\Users\nicho\AppData\Local\Programs\Julia 1.5.0\lib\julia\sys.dll (unknown line)
jl_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\julia.h:1690 [inlined]
start_task at /cygdrive/d/buildbot/worker/package_win64/build/src\task.c:707
Allocations: 806482 (Pool: 806296; Big: 186); GC: 1
```

In Julia 1.4.1, this code results in an exception which does not crash Julia:
```
1-element reinterpret(SalsaBlock, ::Array{UInt128,1}):
Error showing value of type Base.ReinterpretArray{SalsaBlock,1,UInt128,Array{UInt128,1}}:
ERROR: bitcast: argument size does not match size of target type
Stacktrace:
 [1] reinterpret at .\essentials.jl:416 [inlined]
 [2] show(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::SalsaBlock) at .\REPL[2]:1
 [3] sprint(::Function, ::SalsaBlock; context::IOContext{REPL.Terminals.TTYTerminal}, sizehint::Int64) at .\strings\io.jl:103
 [4] alignment at .\show.jl:2063 [inlined]
 [5] alignment(::IOContext{REPL.Terminals.TTYTerminal}, ::Base.ReinterpretArray{SalsaBlock,1,UInt128,Array{UInt128,1}}, ::UnitRange{Int64}, ::UnitRange{Int64}, ::Int64, ::Int64, ::Int64) at .\arrayshow.jl:68
 [6] print_matrix(::IOContext{REPL.Terminals.TTYTerminal}, ::Base.ReinterpretArray{SalsaBlock,1,UInt128,Array{UInt128,1}}, ::String, ::String, ::String, ::String, ::String, ::String, ::Int64, ::Int64) at .\arrayshow.jl:189
 [7] print_matrix at .\arrayshow.jl:162 [inlined]
 [8] print_array at .\arrayshow.jl:310 [inlined]
 [9] show(::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::Base.ReinterpretArray{SalsaBlock,1,UInt128,Array{UInt128,1}}) at .\arrayshow.jl:346
 [10] display(::REPL.REPLDisplay, ::MIME{Symbol("text/plain")}, ::Any) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:137
 [11] display(::REPL.REPLDisplay, ::Any) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:141
 [12] display(::Any) at .\multimedia.jl:323
 [13] #invokelatest#1 at .\essentials.jl:712 [inlined]
 [14] invokelatest at .\essentials.jl:711 [inlined]
 [15] print_response(::IO, ::Any, ::Bool, ::Bool, ::Any) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:161
 [16] print_response(::REPL.AbstractREPL, ::Any, ::Bool, ::Bool) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:146
 [17] (::REPL.var"#do_respond#38"{Bool,REPL.var"#48#57"{REPL.LineEditREPL,REPL.REPLHistoryProvider},REPL.LineEditREPL,REPL.LineEdit.Prompt})(::Any, ::Any, ::Any) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:729
 [18] #invokelatest#1 at .\essentials.jl:712 [inlined]
 [19] invokelatest at .\essentials.jl:711 [inlined]
 [20] run_interface(::REPL.Terminals.TextTerminal, ::REPL.LineEdit.ModalInterface, ::REPL.LineEdit.MIState) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\LineEdit.jl:2354
 [21] run_frontend(::REPL.LineEditREPL, ::REPL.REPLBackendRef) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:1055
 [22] run_repl(::REPL.AbstractREPL, ::Any) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:206
 [23] (::Base.var"#764#766"{Bool,Bool,Bool,Bool})(::Module) at .\client.jl:383
 [24] #invokelatest#1 at .\essentials.jl:712 [inlined]
 [25] invokelatest at .\essentials.jl:711 [inlined]
 [26] run_main_repl(::Bool, ::Bool, ::Bool, ::Bool, ::Bool) at .\client.jl:367
 [27] exec_options(::Base.JLOptions) at .\client.jl:305
 [28] _start() at .\client.jl:484
```

**Comment by jakobnissen** [comment]
Can reproduce on 1.5, but fixed for me on Version 1.6.0-DEV.582.
@BioTurboNick you are indeed doing something very wrong here: `reinterpret(SalsaBlock, zeros(UInt128, 4))` creates a `ReinterpretArray`, not a `SalsaBlock`. Furthermore, `reinterpret(UInt128, ::SalsaBlock)` will not work, because they are of different sizes.

You probably want to do something like this instead:
```
struct SalsaBlock
    data::NTuple{64, UInt8}
end
```
or else look at BitIntegers.jl for how to use large-width primitive types (although LLVM support for large primitives is not great)

**Comment by BioTurboNick** [comment]
Good to know! Then I'll close. And I figured out a way that actually works.
